aspnetcore/fundamentals/dependency-injection.md

```diff
@@ -382,6 +382,34 @@ public class Program
             });
 }
 ```
+## Resolve services in ConfigureService method
+
+In some cases you may need to resolve a service in ConfigureService method, for example, reading a setting from database then decide which service to register. Typically, in asp.net core, the most well known approach to resolve a service is constructor injection. However, in this stage, this approach is not available. So, we can create a service provider in ConfigureService, then resolve the service that we want by utilizing that service provider.
+
+```csharp
+public void ConfigureService(IServiceCollection services)
+{
+    // Configure the services
+    services.AddTransient<ISettingService, DBSettingService>();    
+
+    // Build an intermediate service provider
+    using (var serviceProvider = services.BuildServiceProvider())
+    {
+        // Resolve the services from the service provider
+        var settingService = serviceProvider.GetRequiredService<ISettingService>();
+        // Use your services to decide if FooService and BarService should be injected
+        if(settingService.EnableFoo())
+        {
+            services.AddScoped(IFooService,FooService);
+        }
+        if(settingService.EnableBar())
+        {
+            services.AddScoped(IBarService,BarService);
+        }
```

Comment:
I'd like to try and help suggest an alternative approach for this, but I'm a bit confused by the intent here. If `IFooService` is registered only if `EnableFoo` returns `true`, any services, controllers, etc that depend on `IFooService` are going to see runtime exceptions, because `IFooService` won't have been registered at all. This is similar for `EnableBar` and `IBarService`.

Reply:
@serpent5 Very appericate for your comments, I understand your concern that register different services based on different setting may causes runtime exceptions if a service that not registered is called. And causing multiple singleton instance never comes to my mind before, thanks for point me out through the links. 
This user case may not be best approprate. I can name another senario, here is a sample [thread](https://stackoverflow.com/questions/62050031/using-dependency-injection-to-configure-cookieauthenticationoptions-in-aspnetcor), when configuring CookieAuthentication, developers want to get some settings from database, and that requires a service that registered in DI.
But anyway, if this is a not recommanded approach, it is better that our docs can cover it or we can have a proper solution to get DI instance in configservice method

Reply:
Authentication uses the options pattern, which has built-in support for DI and is covered in the [docs](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-3.1#use-di-services-to-configure-options). Here's an example of how this might look:

```csharp
services.AddAuthentication(CookieAuthenticationDefaults.AuthenticationScheme)
    .AddCookie();

services.AddOptions<CookieAuthenticationOptions>(CookieAuthenticationDefaults.AuthenticationScheme)
    .Configure<IFooService>((options, fooService) =>
    {
        // Use fooService to configure options.
        // ...
    });
```

In this authentication case, the name of the options being configured is the name of the authentication scheme.